UM/Settings/SettingDefinition.py

```diff
@@ -153,7 +153,7 @@ def __setattr__(self, name: str, value: Any) -> None:
     def __hash__(self):
         """Ensure that the SettingDefinition is hashable, so it can be used in a set."""
 
-        return hash(str(self))
+        return hash((id(self), self._key, self._container))
```

Comment:
switching from `str` to `id` can have unforeseen consequence, it will probably be okay in this case, but remember that `id()` is the memory address for the CPython implementation. Speaking of personal experience best to double check how this behaves, when instance are assigned `=`, `copied` or `deepcopied`.

Reply:
I changed all the `__hash__` methods which had a `__repr__` method present, and took the very same used variables, so the result should be stricly similar, without the cost of computing the string.

Reply:
`id()` https://docs.python.org/3/library/functions.html#id is not the same as the dunder `__repr__` https://docs.python.org/3/library/functions.html#repr

Reply:
Btw, I think it should work as expected in this case, but it is my experience that using `id()` can have unwanted side effects because, certainly if assignment `=` (copy of reference) and/or `deepcopy` come into play, not sure if that happens here. I'm just warning that this should be a conscious choice.

Reply:
In general, I agree that using `id()` for the hash is not a good idea, but it depends on whether you want to compare for two objects being the same instance or having the same content.

In this very specific case, the hash was computed on `str(self)`, which actually calls `__repr__` to create the string, then computes the hash of the string. The `__repr__` method of `SettingDefinition` uses fixed characters and the variables `id(self)`, `self._key` and `self._container`, so calling them directly in the hash method is just a shortcut. The computed number will be different, but will evolve the same way, which is what we want for a hash, and gives a similar result to the original code.

Maybe we could still discuss this, because having the ID involved for the hash may be bad indeed. It is useful for debug purposes in the `__repr__` method but may not make sense for the hash. Splitting the string representation and the hash seems to be a good thing to me, because they have different purposes.